src/Aspire.Hosting/ApplicationModel/ResourceLoggerService.cs

```diff
@@ -282,65 +280,44 @@ void Log(InternalLogLine log)
+                lock (_lock)
                 {
-                    logs[i] = new LogLine(lineNumber, entry[i].Message, entry[i].IsError);
-                    lineNumber++;
+                    OnNewLog -= Log;
+                    channel.Writer.TryComplete();
                 }
-
-                return logs;
             }
-        }
 
-        private Task EnsureBacklogReplayAsync(CancellationToken cancellationToken)
-        {
-            lock (_backlog)
+            static LogLine[] CreateLogLines(ref int lineNumber, IReadOnlyList<LogEntry> entries)
             {
-                _backlogReplayCompleteTask ??= StartBacklogReplayAsync(cancellationToken);
-                return _backlogReplayCompleteTask;
-            }
-
-            async Task StartBacklogReplayAsync(CancellationToken cancellationToken)
-            {
-                var delay = TimeSpan.FromMilliseconds(100);
-
-                // There could be an initial burst of logs as they're replayed. Give them the opportunity to be loaded
-                // into the backlog in the correct order and returned before streaming logs as they arrive.
-                for (var i = 0; i < 3; i++)
+                var logs = new LogLine[entries.Count];
+                for (var i = 0; i < entries.Count; i++)
                 {
-                    await Task.Delay(delay, cancellationToken).ConfigureAwait(false);
-                    lock (_backlog)
+                    var entry = entries[i];
+                    var content = entry.Content ?? string.Empty;
+                    if (entry.Timestamp != null)
                     {
-                        if (_lastLogReceivedTimestamp != 0 && Stopwatch.GetElapsedTime(_lastLogReceivedTimestamp) > delay)
-                        {
-                            break;
-                        }
+                        content = entry.Timestamp.Value.ToString(KnownFormats.ConsoleLogsTimestampFormat, CultureInfo.InvariantCulture) + " " + content;
                     }
+
+                    logs[i] = new LogLine(lineNumber, content, entry.Type == LogEntryType.Error);
+                    lineNumber++;
                 }
+
+                return logs;
             }
         }
 
         // This provides the fan out to multiple subscribers.
-        private Action<InternalLogLine>? _onNewLog;
-        private event Action<InternalLogLine> OnNewLog
+        private Action<LogEntry>? _onNewLog;
+        private event Action<LogEntry> OnNewLog
         {
             add
             {
-                bool raiseSubscribersChanged;
-                lock (this)
-                {
-                    raiseSubscribersChanged = _onNewLog is null; // is this the first subscriber?
+                Debug.Assert(Monitor.IsEntered(_lock));
 
-                    _onNewLog += value;
-                }
+                var raiseSubscribersChanged = _onNewLog is null; // is this the first subscriber?
```

Comment:
It's unclear why we only raise a subscriber-change notification for the first subscriber. A comment would be good.